### ISSUE TITLE
--hotswap startup open3 deadlock

### ISSUE THREAD

**Issue by mdkent** [nevans/resque-pool#151]
We've been using the great new --hotswap feature over the last week with generally good results, but this morning we had resque-pool-master fail to start after a deploy, leaving the old revision running. On further inspection the new pool-master died on:

```
/opt/ruby2.3.0/lib/ruby/2.3.0/open3.rb:271:in `value': No live threads left. Deadlock? (fatal)
    from /opt/ruby2.3.0/lib/ruby/2.3.0/open3.rb:271:in `block in capture3'
    from /opt/ruby2.3.0/lib/ruby/2.3.0/open3.rb:205:in `popen_run'
    from /opt/ruby2.3.0/lib/ruby/2.3.0/open3.rb:95:in `popen3'
    from /opt/ruby2.3.0/lib/ruby/2.3.0/open3.rb:258:in `capture3'
    from /u/apps/foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/killer.rb:26:in `all_resque_pool_processes'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/killer.rb:16:in `run'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/killer.rb:11:in `run'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool.rb:93:in `kill_other_pools!'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool.rb:288:in `start'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool.rb:109:in `run'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/tasks.rb:17:in `block (2 levels) in <top (required)>'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:240:in `block in execute'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `each'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:235:in `execute'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:179:in `block in invoke_with_call_chain'
    from /opt/ruby2.3.0/lib/ruby/2.3.0/monitor.rb:214:in `mon_synchronize'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:172:in `invoke_with_call_chain'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/gems/rake-10.5.0/lib/rake/task.rb:165:in `invoke'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/cli.rb:179:in `start_pool'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/lib/resque/pool/cli.rb:21:in `run'
    from /foo/shared/bundle-ruby221/ruby/2.3.0/bundler/gems/resque-pool-0878ccf4e5e9/bin/resque-pool:7:in `<top (required)>'
    from ./bin/resque-pool:3:in `load'
    from ./bin/resque-pool:3:in `<main>'
```

I'm currently trying to reproduce in a standalone snippet but haven't had any luck yet.

**Comment by jeremy** [comment]
`Open3.capture3` is vulnerable to deadlock because it has two threads doing blocking reads on separate stderr/stdout pipes, each with limited buffer size.

In this case, we can simplify to vanilla Kernel#`

#153

**Comment by jeremy** [comment]
@nevans fixed by #153 😊

**Comment by nevans** [comment]
FYI, #153 was merged and released with v0.7.0 (2019-03-08). But I never got around to cleaning up old issues. (Thanks, @jeremy!)
